R/filters.R

```diff
@@ -0,0 +1,57 @@
+#' Main function for data filtering
+#'
+#' @param .data The data to be processed. Must be the list of 2 elements:
+#' data table and metadata table.
+#' @param .method Method of filtering. Implemented methods: by.meta.
+#' @param .query Filtering query. It's a named list of filters that will be applied
+#' to data. Elements of the list are specific for selected method:
+#' - by.meta: names are metadata column headers; each value can be string
+#' (metadata value to select for the column), vector of strings (to select multiple
+#' possible values) or numeric expression in format '>#', '<#', '>=#', '<=#' where
+#' '#' is a number (integer or float, can be negative).
+#'
+#' @examples
+#' data(immdata)
+#'
+#' # Select samples with status "MS"
+#' repFilter(immdata, "by.meta", list(Status = "MS"))
+#'
+#' # Select samples from lanes "A" and "B" with age >= 15
+#' repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">=15"))
+#' @export repFilter
+repFilter <- function(.data, .method, .query, ...) {
+  switch(tolower(.method),
+    by.meta = filter_by_meta(.data, .query),
+    stop("You entered the wrong method! Please, try again.")
+  )
+}
+
+filter_by_meta <- function(.data, .query) {
+  original_classes <- class(.data)
+  names <- names(.query)
+
+  filtered_meta <- .data$meta
+  for (i in seq_along(names)) {
+    name <- names[i]
+    column_query <- unlist(.query[name])
+    filtered_meta <-
+      if (length(column_query) > 1) {
+        filter(filtered_meta, get(name) %in% column_query)
+      } else if (startsWith(column_query, ">=")) {
+        filter(filtered_meta, get(name) >= as.numeric(substring(column_query, 3)))
+      } else if (startsWith(column_query, "<=")) {
+        filter(filtered_meta, get(name) <= as.numeric(substring(column_query, 3)))
+      } else if (startsWith(column_query, ">")) {
+        filter(filtered_meta, get(name) > as.numeric(substring(column_query, 2)))
+      } else if (startsWith(column_query, "<")) {
+        filter(filtered_meta, get(name) < as.numeric(substring(column_query, 2)))
+      } else {
+        filter(filtered_meta, get(name) == column_query)
+      }
```

Comment:
This probably is a good place to raise error if query is not in the right format. In general, consider following simple use cases for error handling:
```repFilter(immdata, "by.meta", list(Status = "MS"))```
- `immdata` doesn't have `$meta`
- `Status` is not in `$meta`
- `"MS"` is not in `Status`
`repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">=15"))`
- repFilter returns empty data because no single row in `$meta` meets required conditions
`repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">*=15"))`
- There is a typo in `.query`

Reply:
There is another one with out-of range:
`repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">=150"))`

Reply:
In general, it looks like there should be unit tests in Immunarch

Reply:
There are unit tests in `tests/testthat`; I want to include simple tests for filters in CALC-102. This edit is about validating user-provided arguments; it can be checked in tests too.

Reply:
Typo `Age = ">*=15"` in query cannot be easily detected, because it's technically valid query to leave only samples with ">*=15" text in column Age. I've added a warning that filter by particular metadata column removed all remaining samples.